code/modules/projectiles/projectile.dm

```diff
@@ -123,23 +137,98 @@
 	return 1 //Bullets don't drift in space
 
 
-/obj/item/projectile/proc/fire()
-	spawn()
-		while(loc)
-			if(kill_count < 1)
-				qdel(src)
-				return
-			if(!paused)
+/obj/item/projectile/proc/fire(var/setAngle)
+	if(setAngle) Angle = setAngle
+	if(!legacy)
+		spawn() //New projectile system
+			while(loc)
+				if(kill_count < 1)
+					qdel(src)
+					return
 				kill_count--
 				if((!( current ) || loc == current))
 					current = locate(Clamp(x+xo,1,world.maxx),Clamp(y+yo,1,world.maxy),z)
-				step_towards(src, current)
-				if((original && original.layer>=2.75) || ismob(original))
+
+				if(!Angle)
+					Angle=round(Get_Angle(src,current))
+				// world << "[Angle] angle"
+				// overlays.Cut()
+				// var/icon/I=new(initial(icon),icon_state) //using initial(icon) makes sure that the angle for that is reset as well
+				// I.Turn(Angle)
+				// I.DrawBox(rgb(255,0,0,50),1,1,32,32)
+				// icon = I
+				if(spread) //Chaotic spread
+					Angle += (rand() - 0.5) * spread
+				var/matrix/M = new//matrix(transform)
+				M.Turn(Angle)
+				transform = M
+
+				var/Pixel_x=round(sin(Angle)+16*sin(Angle)*2)
+				var/Pixel_y=round(cos(Angle)+16*cos(Angle)*2)
+				var/pixel_x_offset = pixel_x + Pixel_x
+				var/pixel_y_offset = pixel_y + Pixel_y
+				var/new_x = x
+				var/new_y = y
+				//Not sure if using whiles for this is good
+				while(pixel_x_offset > 16)
+					// world << "Pre-adjust coords (x++): xy [pixel_x] xy offset [pixel_x_offset]"
+					pixel_x_offset -= 32
+					pixel_x -= 32
+					new_x++// x++
+				while(pixel_x_offset < -16)
+					// world << "Pre-adjust coords (x--): xy [pixel_x] xy offset [pixel_x_offset]"
+					pixel_x_offset += 32
+					pixel_x += 32
+					new_x--
+
+				while(pixel_y_offset > 16)
+					// world << "Pre-adjust coords (y++): py [pixel_y] py offset [pixel_y_offset]"
+					pixel_y_offset -= 32
+					pixel_y -= 32
+					new_y++
+				while(pixel_y_offset < -16)
+					// world << "Pre-adjust coords (y--): py [pixel_y] py offset [pixel_y_offset]"
+					pixel_y_offset += 32
+					pixel_y += 32
+					new_y--
+
+				speed = round(speed) //Just in case.
+				step_towards(src, locate(new_x, new_y, z)) //Original projectiles stepped towards 'current'
+				if(speed <= 1) //We should really only animate at speed 2
+					pixel_x = pixel_x_offset
+					pixel_y = pixel_y_offset
+				else
+					animate(src, pixel_x = pixel_x_offset, pixel_y = pixel_y_offset, time = max(1, (speed <= 3 ? speed - 1 : speed)))
+
+				/*var/turf/T = get_turf(src)
+				if(T)
+					T.color = "#6666FF"
+					spawn(10)
+						T.color = initial(T.color) */
+
+				if(((original && original.layer>=2.75) || ismob(original)))
 					if(loc == get_turf(original))
 						if(!(original in permutated))
-							Bump(original, 1)
-			Range()
-			sleep(1)
+							Bump(original)

```

Comment:
>/obj/item/projectile/Bump(atom/A, yes)
  if(!yes) //prevents double bumps.
    return

that'll probably cause that, yeah.